From 55942b0f6203e97681dd7e94aef54e7d0be4eed6 Mon Sep 17 00:00:00 2001
From: Martin Kampas <martin.kampas@jolla.com>
Date: Mon, 30 Jul 2018 09:33:21 +0200
Subject: [PATCH] Updater: uninstall old components in correct order

If a dependency is removed from a new version of a component, the
updater will uninstall the old components in a random order, ignoring
the dependency existing between the old components.

This is because the stored undo operations are reordered after the
maintenance tool is updated, using the new metadata, where the
dependency does not exist anymore.

This commit fixes this by sorting based on the metadata of the installed
packages, instead of the packages available from online repositories.

With the original approach of sorting operations prior to writing the
configuration, this change would only become effective with the next
maintenance tool update. In order to make the change effective
immediately, operations are reordered unconditionally after loading the
configuration since now.

Task-number: QTIFW-1166
Change-Id: I95abcb6530aae5805b9135d6ee32944ce5095592
---
 src/libs/installer/packagemanagercore_p.cpp | 14 ++++----------
 1 file changed, 4 insertions(+), 10 deletions(-)

diff --git a/src/libs/installer/packagemanagercore_p.cpp b/src/libs/installer/packagemanagercore_p.cpp
index 6bfd73cf..74c65fb9 100644
--- a/src/libs/installer/packagemanagercore_p.cpp
+++ b/src/libs/installer/packagemanagercore_p.cpp
@@ -1332,9 +1332,6 @@ void PackageManagerCorePrivate::writeMaintenanceTool(OperationList performedOper
 #endif
         }
 
-        performedOperations = sortOperationsBasedOnComponentDependencies(performedOperations);
-        m_core->setValue(QLatin1String("installedOperationAreSorted"), QLatin1String("true"));
-
         try {
             QTemporaryFile file;
             QInstaller::openForWrite(&file);
@@ -1641,10 +1638,7 @@ bool PackageManagerCorePrivate::runPackageUpdater()
         QHash<QString, Component *> componentsByName;
 
         // order the operations in the right component dependency order
-        // next loop will save the needed operations in reverse order for uninstallation
-        OperationList performedOperationsOld = m_performedOperationsOld;
-        if (m_core->value(QLatin1String("installedOperationAreSorted")) != QLatin1String("true"))
-            performedOperationsOld = sortOperationsBasedOnComponentDependencies(m_performedOperationsOld);
+        OperationList performedOperationsOld = sortOperationsBasedOnComponentDependencies(m_performedOperationsOld);
 
         // build a list of undo operations based on the checked state of the component
         foreach (Operation *operation, performedOperationsOld) {
@@ -2391,9 +2385,9 @@ OperationList PackageManagerCorePrivate::sortOperationsBasedOnComponentDependenc
 
     const QRegExp dash(QLatin1String("-.*"));
     Graph<QString> componentGraph;  // create the complete component graph
-    foreach (const Component* node, m_core->components(PackageManagerCore::ComponentType::All)) {
-        componentGraph.addNode(node->name());
-        componentGraph.addEdges(node->name(), node->dependencies().replaceInStrings(dash, QString()));
+    foreach (const KDUpdater::LocalPackage &package, localInstalledPackages()) {
+        componentGraph.addNode(package.name);
+        componentGraph.addEdges(package.name, QStringList(package.dependencies).replaceInStrings(dash, QString()));
     }
 
     const QStringList resolvedComponents = componentGraph.sort();
-- 
2.17.1

